pkg/tls/tls_test.go

```diff
@@ -0,0 +1,241 @@
+	template := &x509.Certificate{
+		SerialNumber:          big.NewInt(0),
+		NotBefore:             time.Unix(0, 0),
+		NotAfter:              time.Unix(0, 0),
+		Subject:               pkix.Name{CommonName: cn, Organization: orgs},
+		Issuer:                pkix.Name{CommonName: cn},
+		IPAddresses:           []net.IP{net.ParseIP("192.168.0.1")},
+		DNSNames:              []string{cn},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth | x509.ExtKeyUsageClientAuth},
+		IsCA:                  isCA,
+	}
+	// let's make a key real quick
+	pk, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Errorf("Error generating temp private key %s", err)
+	}
+	knownGoodDER, err := x509.CreateCertificate(rand.Reader, template, template, pk.Public(), pk)
+	if err != nil {
+		fmt.Errorf("Error generating known good cert %s", err)
+	}
+	knownGood, err := x509.ParseCertificate(knownGoodDER)
+	if err != nil {
+		fmt.Errorf("unable to parse DER %s", err)
+	}
+	return knownGood
+}
+
+func loadGoodCert(path string) *x509.Certificate {
+
+	knownGoodBlob, err := ioutil.ReadFile(path)
+	if err != nil {
+		fmt.Errorf("unable to load last known good cert from disk %s", err)
+	}
+	block, _ := pem.Decode(knownGoodBlob)
+	if block == nil || block.Type != "CERTIFICATE" {
+		fmt.Errorf("failed to decode PEM block containing public key")
+	}
+
+	knownGoodCerts, err := x509.ParseCertificates(block.Bytes)
+	if err != nil {
+		fmt.Errorf("unable to parse certificates %s", err)
+	}
+	goodCert := knownGoodCerts[0]
+	return goodCert
+}
+
+const cn = "dummy-test-certificate.local"
+const knownGoodLoc = "./testdata/known_good_cert.pem"
+const knownGoodLocCA = "./testdata/known_good_certCA.pem"
+const updateSSLVar = "UPDATE_KNOWN_SSL_CERT"
+
+var orgs = []string{cn}
+
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	if newPK.Validate() != nil {
+		t.Errorf("Failed to validate private key")
+	}
+	// the exponent which is effectively the length is encoded inside PublicKey
+	keyLength := newPK.PublicKey.N.BitLen()
+	if keyLength < 2048 {
+		t.Errorf("insecure key length detected: %d", keyLength)
+	}
+}
+
+func TestNewCA(t *testing.T) {
+	pk, cert, err := NewCA(cn)
+	if err != nil {
+		t.Errorf("failed to create new CA %s", err)
+	}
+	if pk.Validate() != nil {
+		t.Errorf("failed to validate RSA Private key")
+	}
+	if os.Getenv(updateSSLVar) == "true" {
+		// reuse functionality from marshal.go
+		pemEncodedCert, err := CertAsBytes(createKnownGoodCert(true))
+		if err != nil {
+			t.Errorf("Cannot PEM encode certificate")
+		}
+		err = ioutil.WriteFile(knownGoodLocCA, pemEncodedCert, 0644)
+		if err != nil {
+			t.Errorf("unable to save PEM cert %s", err)
+		}
+	}
+	goodCert := loadGoodCert(knownGoodLocCA)
+	cert.SerialNumber = big.NewInt(0)
```

Comment:
any pointer or slice fields that need to be nilled - set them to nil.  Others - set them directly to the public struct zero value